### DISCUSSION TITLE
Allowing metadata to be specified separately from media files, AKA sidecars / tagfiles

### DISCUSSION THREAD

**Discussion by mgdigital** [navidrome/navidrome#3181] [Ideas]
It's often desirable to retain the untouched media files when tags need to be modified. Currently, this means keeping 2 copies of the files.

Prior work done in this area is m-Tags, which are/were supported by a Foobar2000 plugin. Here is an archive of the now-defunct website: http://web.archive.org/web/20190915154126/http://m-tags.org/index.html - It's simply a JSON file that sits alongside the media files and allows tags to be overridden.

I'd find something like this very useful - anyone else?

**Comment by 29821632** [comment]
I saw your work on the PR (https://github.com/navidrome/navidrome/pull/3312) a few months ago and I like the approach.
Navidrome is seriously awesome software but unfortunately it doesn't quite fit my use-case of having a completely untagged library (the metadata of my library is defined in directory and file naming so duplicating and syncing that information into the file metadata simply never made sense to me).
Untagged libraries will not be supported directly in Navidrome which is totally fair enough (https://www.navidrome.org/docs/faq/#-can-you-add-a-browsing-by-folder-optionmode-to-navidrome).

So my intention (https://github.com/navidrome/navidrome/issues/560) was to create a fallback method of inferring metadata when Navidrome encounters an untagged file but I never found the time to look into it (I rarely work in Golang with any intensity so it takes me a while to grok unfamiliar codebases). Your PR has given me inspiration and the entry points on how to get moving on my idea... thanks!

**Comment by mgdigital** [comment]
Okay, I may be going it alone with this, but I've been using my tagfiles/sidecar implementation for months now, and since my initial draft PR (https://github.com/navidrome/navidrome/pull/3312) have enhanced the implementation and merged in the BFR updates from `master`.

My implementation is currently on the `tagfiles-bfr` branch here: https://github.com/mgdigital/navidrome
Compare my changes to current Navidrome `master` branch: https://github.com/navidrome/navidrome/compare/0d42b9a4a5436f65142b97ba2619f411219785e7...mgdigital:navidrome:tagfiles-bfr

I'm closing my previous draft PR for now. I fully understand that not everyone will want this and that the implementation introduces a new standard along with a maintenance burden, but @deluan will perhaps be interested to take a look anyway.

## Examples

**1. An artist compilation has not been tagged with `albumartist`**, and so is appearing as a separate album for each track.

**Solution:** place the following `tags.yml` file in the album's folder:

```yaml
albumartist: DJ Yoda
```

**2. A multi-disc album has a different album tag for each disc**, and has been placed in subfolders `CD1` and `CD2`.

**Solution:** place the following `tags.yml` file in the album's root folder. This file uses the syntax of a top level glob pattern, with overridden key values for each pattern:

```yaml
"CD*/*.flac":
  album: Physical Graffiti
  totaldiscs: 2
"CD1/*.flac":
  discnumber: 1
"CD2/*.flac":
  discnumber: 2
```

**3. Albums by the same artist have variations of the artist name** (e.g. `alt-J` and `Alt J`), leading to duplicate entries for the same artist. Some albums have been tagged with `album artist`, and others with `albumartist`.

**Solution:** place the following `tags.yml` file in the root folder for the artist, to set `albumartist` on all files in all subfolders, and remove any `album artist` tags:

```yaml
albumartist: Whatever
"album artist": null
```

### NEW: Arbitrary tag transformations using CEL expressions

Media files from disparate sources can sometimes have tagging anomalies that would be labourious to fix using the above methods. I therefore added the possibility to use [CEL expressions](https://cel.dev/) to apply scripted tag transformations.

A key of `$` under a tag name indicates that the value is a CEL expression.

Here is just 1 example that I recently dealt with. It concerns a DJ mix where the DJ name has been put in the `artist` tag of every track, with the `title` tag containing a concatenation of the actual artist, for example:

| artist   | title |
| -------- | ------- |
| Name of DJ  | Dolly Parton & Kenny Rogers - Islands in the Stream    |
| Name of DJ | Lipps Inc. - Funkytown    |

This can be fixed with the following `tags.yml` file:

```yaml
albumartist: Name of DJ
artist:
  $: tags["title"][0].split(" - ")[0].split(" & ")
title:
  $: value.split(" - ")[1]
```

You'll then end up with artists split correctly, and the title no longer containing the artist name.

This can be very flexible, for example it could be used for path-and-filename-base tag inference, which although I wouldn't personally want it has been requested a few times. Here's an example of this:

https://github.com/mgdigital/navidrome/blob/38efb691a3643f7a1c9640bacfe33549e6236ef3/scanner/metafiles/parse_tags_file_test.go#L147-L169

If there's interest I can provide more documentation on the expressions - for now you'd have to look at my code and the CEL docs to work out what's possible with this.

---

@29821632 Thanks for being the only person to respond on this discussion, and please see my comment about about path-and-filename-base tag inference - perhaps you could become the second user of my fork!?

>  it doesn't quite fit my use-case of having a completely untagged library (the metadata of my library is defined in directory and file naming so duplicating and syncing that information into the file metadata simply never made sense to me).

The problem with this is that the directory/filename can't always encode all the necessary info; for example some characters are not allowed in directory/folder names. But either way it would be possible with my implementation.

**Comment by mist2t** [comment]
I think this is cool. It's also worth having some type of sidecar system that allows other metadata that it's not stored in files to reside outside the DB, acting also like a backup. On each scan / re-scan those sidecar files can be read and imported into DB.

One particular scenario I would use this is for artist bio and artist genre, album description etc. It's nice to have a sidecar file for each artist and each album defining this "special" metadata.

**Reply by mgdigital** [reply]
With this implementation I can easily set genres to all tracks at the artist level because my folder structure is organised by artist, so I simply put a `tags.yml` in the root artist folder. You might then define further genres at the per-track or per-album level, which is also possible here. That's where it differs from a bio or album description because there's no tag supporting these (unless you shoehorn something into the comment field) and it wouldn't make sense to have a per-track artist bio anyway. 

So yes I agree that sidecars could provide a bio or album description too, but I think it would be a separate feature from the `tags.yml` feature, which is specifically for correcting tags on media files.

**Reply by mist2t** [reply]
Oh, I get it now ...
Your solution provides an "external" metadata tag "system" for files that don't have embedded tags ... or can be used as a "master override" system for whatever tags might be in the files themselves.

It is cool indeed. 

Maybe it can be used to fulfill all those 3 scenarios:

1. Provide "master" tags for un-tagged files
2. Provide "master" tags override for whatever tags might be in the files.
3. Provide support for "special" metadata used in the UI: artists bio, albums description etc.

Also since it's manual work, this might be super well documented to describe all the case scenarios and proper formatting on these yml files + their placement.

I was suggesting Kodi NFO "templates" for artist and albums (https://kodi.wiki/view/NFO_files/Templates)  because it is widely used in Jellyfin and Emby (maybe in other media servers too) and some users already might have these files in place. 
As you said, maybe these XML nfo's are pretty limited for the functionality you want to achieve.

**Reply by mgdigital** [reply]
Yep. My primary use case is number 2. Tags found on media files are often wrong (or at least not optimal for Navidrome) and there are, *ahem, good reasons*, not to alter the content of the media files themselves.

I did look at Kodi nfo files but yes, they don't allow me to achieve what I want, aside from which the XML format is horrible to work with!

Agreed that if the idea gains traction then the format would need better documenting and maybe some tooling created for generating the files more easily.

**Comment by lc-guy** [comment]
I've been looking for something like this for a while, so it's nice to see it implemented here.

**Comment by michaelkrieger** [comment]
Referencing my two use cases at https://github.com/navidrome/navidrome/discussions/2853#discussion-6204808 and https://github.com/navidrome/navidrome/discussions/2853#discussioncomment-9205622

this is a great contribution.

**Comment by surskitt** [comment]
I have those same good reasons to not modify the source music files, so something like this would be extremely useful for me.

**Comment by troggy** [comment]
@mgdigital appreciate your effort on making this possible. I think I'm going to try out your branch (really need to keep untagged files). 

1. Just curious what was your rationale not to implement m-tags? I understand that YAML is much terser than JSON, but still?
2. How does tagging on track level look like? Can I have single tags.yml per album and specify all the album and track metadata there?

I guess I just miss a spec or a real-life example of tags.yml for a fully untagged album

**Comment by ChilledSlim** [comment]
Would love to see something implemented for this.

**Comment by nicolas-lesage** [comment]
Me too!

A few thoughts:

- I've been using immich for a few months for my photo library, where sidecar files that contain exif information are taken into account. For each picture, e.g. `IMG_001.jpg`, a sidecar file is created: `IMG_001.jpg.xmp`, in order not to modify original files when changing the picture metadata. It would be nice to have a similar implementation with Navidrome, for example if I have a `01 Awesome Song.m4a` in my library, I could add information by creating a `01 Awesome Song.m4a.yml` (or .xmp / .nfo / .whateverIsRelevant).

- My use case is especially for songs with multiple artists. I'm migrating from 4 iTunes libraries (supposed to be maintained and have similar songs, which was a nightmare) to Navidrome on a NAS. I've created a migration script to import the play counts and last play date from iTunes library to Navidrome database, that I will run 4 times (one with each iTunes library). Hence the need to not modify source files. However, some song have multiple artists (e.g. `ArtistX Feat. ArtistY & ArtistZ`, or `ArtistA x ArtistB`), that are not correctly recognized by Navidrome (and I cannot add `&` or `x` to the recognition patern, since it could mess up artists like "Angus & Julia Stone"). As a result, having the possibility to add the following sidecar file to relevant song files would be awesome:

```
Artist: ArtistX Feat. ArtistY & ArtistZ
Artists:
- ArtistX
- ArtistY
- ArtistZ
```

- I also checked pluggins since there is [an existing pluggin about artist/album metadata](https://github.com/metalheim/navidrome-plugin-artist-nfo-metadata), but it seems that "with the current documented plugin API, a plugin can probably read sidecar files, but it cannot make Navidrome use them as track scan metadata" (I'm quoting my favorite AI chatbot here).

I don't know Go but I could try to learn and would be happy to help. I'd like to avoid using a fork since it would become difficult to maintain it (Docker with latest Navidrome version on the NAS is of course easier).

**Comment by graynk** [comment]
Just chiming in that I'd love to see this implemented too
